pkg/crcontrollers/storagegroup/storagegroupcontroller.go

```diff
@@ -0,0 +1,586 @@
+	if err := c.client.ReadList(ctx, drivesList); err != nil {
+		log.Errorf("failed to read drives list: %s", err.Error())
+		return ctrl.Result{Requeue: true}, err
+	}
+	labelingNoError := true
+	noDriveSelected := true
+	drivesCount := map[string]int32{}
+	driveSelector := sg.Spec.DriveSelector
+
+	// used for candidate drives to be selected by storagegroup with numberDrivesPerNode > 0
+	var candidateDrives []*drivecrd.Drive
+	for _, d := range drivesList.Items {
+		drive := d
+		existingStorageGroup, exists := drive.Labels[apiV1.StorageGroupLabelKey]
+		if exists {
+			if existingStorageGroup == sg.Name {
+				log.Debugf("Drive %s has already been selected by current storage group", drive.Name)
+				noDriveSelected = false
+				if driveSelector.NumberDrivesPerNode > 0 {
+					drivesCount[drive.Spec.NodeId]++
+				}
+			}
+			continue
+		}
+
+		if isDriveSelectedByValidMatchFields(&drive.Spec, &driveSelector.MatchFields) &&
+			(driveSelector.NumberDrivesPerNode == 0 || drivesCount[drive.Spec.NodeId] < driveSelector.NumberDrivesPerNode) {
+			if driveSelector.NumberDrivesPerNode > 0 {
+				candidateDrives = append(candidateDrives, &drive)
+				continue
+			}
+
+			if err := c.addDriveAndACStorageGroupLabel(ctx, log, &drive, sg); err != nil {
+				log.Errorf("Error in adding storage-group label to drive %s: %s", drive.Name, err.Error())
+				labelingNoError = false
+			}
+			noDriveSelected = false
+		}
+	}
+
+	for _, d := range candidateDrives {
+		drive := d
+		if drivesCount[drive.Spec.NodeId] < driveSelector.NumberDrivesPerNode {
+			if err := c.addDriveAndACStorageGroupLabel(ctx, log, drive, sg); err != nil {
+				log.Errorf("Error in adding storage-group label to drive %s: %s", drive.Name, err.Error())
+				labelingNoError = false
+			}
+			noDriveSelected = false
+			drivesCount[drive.Spec.NodeId]++
+		}
+	}
+
+	if noDriveSelected {
+		log.Warnf("No drive can be selected by current storage group %s", sg.Name)
+	}
+	if labelingNoError {
+		sg.Annotations[sgTempStatusAnnotationKey] = apiV1.Created
+		if err := c.client.UpdateCR(ctx, sg); err != nil {
+			log.Errorf("Unable to update StorageGroup status, error: %v.", err)
+			return ctrl.Result{Requeue: true}, err
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{Requeue: true}, fmt.Errorf("error in adding storage-group label")
+}
+
+func isDriveSelectedByValidMatchFields(drive *api.Drive, matchFields *map[string]string) bool {
+	for fieldName, fieldValue := range *matchFields {
+		driveField := reflect.ValueOf(drive).Elem().FieldByName(fieldName)
+		switch driveField.Type().String() {
+		case "string":
+			if driveField.String() != fieldValue {
+				return false
+			}
+		case "int64":
+			fieldValueInt64, _ := strconv.ParseInt(fieldValue, 10, 64)
+			if driveField.Int() != fieldValueInt64 {
+				return false
+			}
+		case "bool":
+			fieldValueBool, _ := strconv.ParseBool(fieldValue)
+			if driveField.Bool() != fieldValueBool {
+				return false
+			}
```

Comment:
here should we add default case here for un-intention type ?

Reply:
I enumerate all the drive CR's property types here. Ok, I will change the code to add default case for un-intention type according to your comment.